src/lib.rs

```diff
@@ -149,23 +149,92 @@ impl Key for &str {
-/// by all implementations in this crate. It is only violated
+/// # Design
+///
+/// The const parameter `Y` is the miniconf recursion depth. It defaults to `1`.
+///
+/// An implementor of `Miniconf<Y>` may consume at most `Y` items from the
+/// `keys` iterator argument in the recursive methods ([`Miniconf::serialize_by_key()`],
+/// [`Miniconf::deserialize_by_key()`], [`Miniconf::traverse_by_key()`]). This includes
+/// both the items consumed directly before recursing and those consumed indirectly
+/// by recursing into inner types. In the same way it may call `func` in
+/// [`Miniconf::traverse_by_key()`] at most `Y` times, again including those calls due
+/// to recursion into inner `Miniconf` types.
+///
+/// This implies that if an implementor `T` of `Miniconf<Y>` contains and recurses into
+/// an inner type using that type's `Miniconf<Z>` implementation, then `Z <= Y` must
+/// hold and `T` may consume at most `Y - Z` items from the `keys` iterators and call
+/// `func` at most `Y - Z` times.
+///
+/// The recursion depth `Y` is thus an upper bound of the maximum key length
+/// (the depth/height of the tree).
+///
+/// # Derive macro
+///
+/// A derive macro to automatically implement the correct `Miniconf<Y>` on a struct `S` is available at
+/// [`macro@Miniconf`].
+///
+/// Each field in the struct must either implement [`serde::Serialize`] `+` [`serde::de::DeserializeOwned`]
+/// (and be supported by the intended [`serde::Serializer`]/[`serde::Deserializer`] backend)
+/// or implement [Miniconf].
+///
+/// For each field, the Miniconf recursion depth is configured through the `#[miniconf(defer(Y))]` attribute,
+/// with `Y = 1` being the implied default when using `#[miniconf(defer)]` and `Y = 0` invalid.
+/// If the attribute is not present, the field is a leaf and accessed only through its
+/// [`serde::Serialize`]/[`serde::Deserialize`] implementation.
+/// With the attribute present the field is accessed through its [`Miniconf<Y>`] implementation with the given
+/// recursion depth.
+///
+/// Homogeneous [core::array]s can be made accessible either
+/// 1. as a single leaf in the tree like other serde-capable items, or
+/// 2. by item through their numeric indices (with the attribute `#[miniconf(defer(1))]`), or
+/// 3. exposing a sub-tree per item with `#[miniconf(defer(D))]` and `D >= 2`.
+///
+/// `Option` is used
+/// 1. as a leaf like a standard `serde` Option, or
+/// 2. with `#[miniconf(defer(1))]` to support a leaf value that may be absent (masked) at runtime.
+/// 3. with `#[miniconf(defer(D))]` and `D >= 2` to support masking sub-trees at runtime.
+///
+/// ## Bounds on generics
+///
+/// The macro adds bounds to generic types of the struct it is acting on.
+/// E.g. If a generic type parameter `T` of the struct `S<T>`is used as a type parameter to a
+/// field type `a: F1<F2<T>>` the type `T` will be considered to reside at depth `X = 2` (as it is
```

Comment:
Can we just write this out as:
```rs
type F1<T> = ...;
type F2<T> = ...;

struct S<T> {
    a: F1<F2<T>>,
}
```

It actually took me a couple read throughs to realize this is what it was trying to describe.

Reply:
Added doctest.